src/common/enumiterator.h

```diff
@@ -0,0 +1,155 @@
+     * of size_t or ssize_t depending on the sign of the value_type.
+     *
+     * Value of stop equal to selected MAX is not allowed!
+     *
+     * @{ */
+    using stored_type =
+            std::common_type_t<
+                raw_type
+              , std::conditional_t<
+                    std::is_signed<raw_type>::value
+                  , std::ptrdiff_t // FIXME: ssize_t?
+                  , std::size_t
+                  >
+              >;
+
+    static constexpr stored_type makeEnd()
+    {
+        return std::numeric_limits<stored_type>::max();
+    }
+    /* @} */
+
+ public:
+
+    /** Creates a singular iterator. */
+    EnumIterator() = default;
+
+    explicit EnumIterator(T t)
+      : _current{static_cast<stored_type>(t)}
+    {}
+
+    // Iterator concept requirements {
+    using difference_type   = std::ptrdiff_t;
+    using value_type        = T;
+    using pointer           = std::add_pointer_t<value_type>;
+    using reference         = std::add_lvalue_reference_t<value_type>;
+    using iterator_category = std::forward_iterator_tag;
+
+    // extra two:
+    using const_reference   = std::add_lvalue_reference_t<std::add_const_t<value_type>>;
+    using const_pointer     = std::add_pointer_t<std::add_const_t<value_type>>;
+
+    EnumIterator(const EnumIterator&)            = default;
+    EnumIterator(EnumIterator&&)                 = default;
+    EnumIterator& operator=(const EnumIterator&) = default;
+    EnumIterator& operator=(EnumIterator&&)      = default;
+    ~EnumIterator()                              = default;
+
+    friend void swap(EnumIterator& lhs, EnumIterator& rhs)
+    {
+        std::swap(lhs._current, rhs._current);
+    }
+    // }
+
+    // EqualityComparable concept -- no floating points {
+    bool operator==(EnumIterator rhs) const { return _current == rhs._current; }
+    bool operator!=(EnumIterator rhs) const { return ! (*this == rhs);         }
+    // }
+
+    // InputIterator concept {
+    EnumIterator& operator++()
+    {
+        if (makeEnd() != _current)  // no point to advance if at the end
+        {
+            if (_current < static_cast<stored_type>(stop))
+            {
+                ++_current;
+            }
+            else
+            {
+                _current = makeEnd();
+            }
+        }
+
+        return *this;
+    }
+
+    EnumIterator operator++(int) const
+    {
+        return (makeEnd() != _current) ? ++EnumIterator{*this} : EnumIterator{};
+    }
+    // }
+
+    EnumIterator begin() const { return EnumIterator{start}; }
+    EnumIterator end()   const { return {};      } // value-initialised singular is past-the-end
+
+    reference operator*()
+    {
+        assert(derefable());
+
+        return *reinterpret_cast<pointer>(&_current);  // FIXME
+    }
+
+    const_reference operator*() const
+    {
+        assert(derefable());
+
+        return *reinterpret_cast<const_pointer>(&_current); // FIXME
```

Comment:
Why do we have fixme here? Can this be fixed before PR is merged?

Reply:
Ditto.